### PR TITLE
add  "Deny device creation via gateway" option for gateway

### PR DESCRIPTION
**Pull request by liaow**
- gateway can not create device on connect if name not exists when "Deny device creation via gateway" option checked.
- gateway can create device on connect if name not exists for default .
- to define a gateway device, tenant user need to check "is gateway" and "Deny device creation via gateway" option.

wx20171024-110305 2x

### CODE REVIEW

**Comment by ashvayka** [comment]
@liaow Thank you. I will review this and will get back to you with some tips to improve performance and error handling. stay tuned

**Comment by ssmaurya** [comment]
Few more things to be added: filter gateway devices so that we could see all the devices from particular gateway. Plus add option for allowing or blocking every gateway device.

**Comment by ssmaurya** [comment]
Default for gateway devices could be allowed.

**Comment by liaow** [comment]
@ssmaurya How should I deal these details?

File: transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/session/GatewaySessionCtx.java [hunk 1]
```diff
@@ -89,6 +90,9 @@ private void onDeviceConnect(String deviceName, String deviceType) {
         if (!devices.containsKey(deviceName)) {
             Optional<Device> deviceOpt = deviceService.findDeviceByTenantIdAndName(gateway.getTenantId(), deviceName);
             Device device = deviceOpt.orElseGet(() -> {
+                JsonNode forbidCreateDevice = gateway.getAdditionalInfo().get("forbid_create_device");
```
Comment (ashvayka) [inline transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/session/GatewaySessionCtx.java]: please rename to "allowDeviceCreation" to be camel case. (The boolean condition need to be reversed).

File: transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/session/GatewaySessionCtx.java [hunk 2]
```diff
@@ -89,6 +90,9 @@ private void onDeviceConnect(String deviceName, String deviceType) {
...
+                if (forbidCreateDevice != null && forbidCreateDevice.asBoolean())
```
Comment (ashvayka) [inline transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/session/GatewaySessionCtx.java]: please use bracers {} in if statements and in blocks

File: transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/session/GatewaySessionCtx.java (hunk 1 above)
Comment (ashvayka) [reply transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/session/GatewaySessionCtx.java]: This boolean flag should be fetched. No need to get and calculate this each time the device sends a request.

File: transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/session/GatewaySessionCtx.java [hunk 3]
```diff
@@ -97,6 +101,10 @@ private void onDeviceConnect(String deviceName, String deviceType) {
                 relationService.saveRelationAsync(new EntityRelation(gateway.getId(), newDevice.getId(), "Created"));
                 return newDevice;
             });
+            if (null == device) {
+                log.debug("[{}] NOT allowed add to add device [{}] to the gateway session", gatewaySessionId, deviceName);
```
Comment (ashvayka) [inline transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/session/GatewaySessionCtx.java]: please change this to warning. Also, the data upload request will be failed, but gateway can't know if it is doing something wrong. So, the best idea is to throw RuntimeException or AdaptorException and close gateway session in the MqttTransportHandler.handleMqttPublishMsg

File: transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/session/GatewaySessionCtx.java (hunk 1 above)
Comment (liaow) [reply transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/session/GatewaySessionCtx.java]: sorry, it is hard to find a suitable place to set this attribute for gateway device.

**Pull request closed without merging**